{{cookiecutter.repo_name}}/src/{{cookiecutter.repo_name_slug}}/{{cookiecutter.conversion_name}}/{{cookiecutter.conversion_name}}_convert_all_sessions.py

```diff
@@ -0,0 +1,120 @@
+"""Primary script to run to convert all sessions in a dataset using session_to_nwb."""
+from pathlib import Path
+from typing import Union
+from concurrent.futures import ProcessPoolExecutor, as_completed
+from pprint import pformat
+import traceback
+from tqdm import tqdm
+
+from .{{cookiecutter.conversion_name}}_convert_session import session_to_nwb
+
+
+def dataset_to_nwb(
+    *,
+    data_dir_path: Union[str, Path],
+    output_dir_path: Union[str, Path],
+    max_workers: int = 1,
+    stub_test: bool = False,
+    verbose: bool = True,
+):
+    """Convert the entire dataset to NWB.
+
+    Parameters
+    ----------
+    data_dir_path : Union[str, Path]
+        The path to the directory containing the raw data.
+    output_dir_path : Union[str, Path]
+        The path to the directory where the NWB files will be saved.
+    max_workers : int, optional
+        The number of workers to use for parallel processing, by default 1
+    stub_test : bool, optional
+        Whether to run a stub test, by default False
+    verbose : bool, optional
+        Whether to print verbose output, by default True
+    """
+    data_dir_path = Path(data_dir_path)
+    session_to_nwb_kwargs_per_session = get_session_to_nwb_kwargs_per_session(
+        data_dir_path=data_dir_path,
+        output_dir_path=output_dir_path,
+        stub_test=stub_test,
+        verbose=verbose,
+    )
+
+    futures = []
+    with ProcessPoolExecutor(max_workers=max_workers) as executor:
+        for session_to_nwb_kwargs in session_to_nwb_kwargs_per_session:
+            exception_file_path = data_dir_path / f"ERROR_<nwbfile_name>.txt" # Add error file path here
+            futures.append(
+                executor.submit(
+                    safe_session_to_nwb,
+                    session_to_nwb_kwargs=session_to_nwb_kwargs,
+                    exception_file_path=exception_file_path,
+                )
+            )
+        for _ in tqdm(as_completed(futures), total=len(futures)):
+            pass
+
+
+def safe_session_to_nwb(*, session_to_nwb_kwargs: dict, exception_file_path: Union[Path, str]):
+    """Convert a session to NWB while handling any errors by recording error messages to the exception_file_path.
+
+    Parameters
+    ----------
+    session_to_nwb_kwargs : dict
+        The arguments for session_to_nwb.
+    exception_file_path : Path
+        The path to the file where the exception messages will be saved.
+    """
+    exception_file_path = Path(exception_file_path)
+    try:
+        session_to_nwb(**session_to_nwb_kwargs)
+    except Exception as e:
+        with open(exception_file_path, mode="w") as f:
+            f.write(f"session_to_nwb_kwargs: \n {pformat(session_to_nwb_kwargs)}\n\n")
+            f.write(traceback.format_exc())
+
+
+def get_session_to_nwb_kwargs_per_session(
+    *,
+    data_dir_path: Union[str, Path],
+    output_dir_path: Union[str, Path],
+    stub_test: bool = False,
```

Comment:
Why would this take stub_test as a keyword? I feel it makes sense I Just want to think on a concrete example.

Reply:
tbh, I never use stub_test anyway, so I wouldn't know. I included it here just to match the signature of `session_to_nwb`.  If you don't think it makes sense here, we can remove it just lmk.

Reply:
I was curios on whether you had some specific use case on mind, I also can't think on a use case for the output_dir_path here.

Yeah, let's spend additional words on the purpose of the function and remove the parameters that users of this might have to remove:
* Verbose
* output_dir_path
* stub_test

Reply:
Oh I must've misunderstood a bit.  `stub_test` probably doesn't need to be passed along, but the others should. These other parameters are keyword arguments for session_to_nwb, so they need to be included in `session_to_nwb_kwargs_per_session` for the whole script to work properly.

Reply:
Maybe this simple example implementation will help show what I'm talking about here.
```python
def get_session_to_nwb_kwargs_per_session(
    *,
    data_dir_path: Union[str, Path],
    output_dir_path: Union[str, Path],
    verbose: bool = True,
):
    """Get the kwargs for session_to_nwb for each session in the dataset.

    Parameters
    ----------
    data_dir_path : Union[str, Path]
        The path to the directory containing the raw data.
    output_dir_path : Union[str, Path]
        The path to the directory where the NWB files will be saved.
    verbose : bool, optional
        Whether to print verbose output, by default True

    Returns
    -------
    list[dict[str, Any]]
        A list of dictionaries containing the kwargs for session_to_nwb for each session.
    """
    data_dir_path = Path(data_dir_path)
    session_to_nwb_kwargs_per_session = []
    for data_file_path in data_dir_path.iterdir():
        session_to_nwb_kwargs = dict(
            data_file_path=data_file_path, # suppose that session_to_nwb takes data_file_path as an argument
            output_dir_path=output_dir_path,
            verbose=verbose,
        )
        session_to_nwb_kwargs_per_session.append(session_to_nwb_kwargs)
    return session_to_nwb_kwargs_per_session

```

Reply:
There are kwargs that change per session. I think that `data_file_path` would be the on in your example. My understanding is that `get_session_to_nwb_kargs_per_session` allows you to determine that.  But `verbose` and `output_dir` do not change per session, they are global so I don't think they need to be here.  In other words, why make the function more complicated if they can just be set at `session_to_nwb`:

```python

session_to_nwb(output_dir=output_dir, verbose=verbose, stub_test=stub_test, **session_to_nwb_kwargs_per_session)

```

Am I missing something?

Reply:
That makes sense to me.